src/glview/UIGeometryRenderer.cc

```diff
@@ -0,0 +1,99 @@
+#include "UIGeometryRenderer.h"
+
+#include "glview/PolySetRenderer.h"
+#include "core/ColorNode.h"
+#include "core/ModuleInstantiation.h"
+#include "core/TransformNode.h"
+#include "core/Tree.h"
+#include "core/primitives.h"
+#include "geometry/GeometryEvaluator.h"
+
+UIGeometryRenderer::UIGeometryRenderer(double zoomValue)
+  : zoomValue(zoomValue)
+  , module(new ModuleInstantiation("UI"))
+  , root(new GroupNode(module.get()))
+{
+
+}
+
+void UIGeometryRenderer::addPoint(const Vector3d& pt, const Color4f& color)
+{
+  auto tr = std::make_shared<TransformNode>(module.get(), "multmatrix");
+  tr->matrix = Eigen::Translation3d(pt);
+  tr->children.push_back(makePointSphereNode());
+  root->children.push_back(makeColorNode(tr, color));
+}
+
+void UIGeometryRenderer::addLine(const Vector3d& start, const Vector3d& end, const Color4f& color)
+{
+  double len = (end - start).norm();
+
+  auto tr1 = std::make_shared<TransformNode>(module.get(), "multmatrix");
+  tr1->matrix = Eigen::Translation3d(start) * Eigen::Quaterniond::FromTwoVectors(Vector3d::UnitZ(), end - start);
+
+  tr1->children.push_back(makePointSphereNode());
+
+  auto tr2 = std::make_shared<TransformNode>(module.get(), "multmatrix");
+  tr2->matrix = Eigen::Translation3d(Vector3d(0.0, 0.0, len));
+  tr2->children.push_back(makePointSphereNode());
+
+  tr1->children.push_back(tr2);
+
+  // This magic factor makes the cylinder and sphere radii match up closely
+  double r = zoomValue * 0.002 * 0.92;
+  auto cyl = std::make_shared<CylinderNode>(module.get());
+  cyl->fn = 8;
+  cyl->r1 = r;
+  cyl->r2 = r;
+  cyl->h = len;
+  tr1->children.push_back(cyl);
+
+  root->children.push_back(makeColorNode(tr1, color));
+}
+
+void UIGeometryRenderer::draw(ShaderUtils::ShaderInfo *shaderInfo)
+{
+  if (root->children.empty()) {
+    return;
+  }
+
+  Tree tree(root, "GUI");
+  GeometryEvaluator evaluator(tree);
+  std::shared_ptr<const Geometry> geometry = evaluator.evaluateGeometry(*tree.root(), true);
+
+  PolySetRenderer renderer(geometry);
+  renderer.prepare(shaderInfo);
```

Comment:
While this is general, it feels slightly wasteful to go through an entire geometry evaluation per rendered frame.
I guess this depends a bit on the use-case; interactive selection while the object is static vs. moving the camera once the markup is rendered. My worry is mostly about performance behavior of the latter case.

Reply:
I didn't make anything persist across frames because, at least for now, the amount of geometry is pretty small and only evaluated when the measurement tool is in use.

If you think that's still too much of a performance hit, I can add some caching.

Reply:
It might pan out - I need to think about it and test it. I'm just a bit surprised about evaluating CSG geometry per frame for rendering annotations :)